drivers/unix/os_unix.cpp

```diff
@@ -720,14 +732,47 @@ int OS_Unix::get_process_id() const {
 }
 
 bool OS_Unix::is_process_running(const ProcessID &p_pid) const {
+	process_map_mutex.lock();
+	const ProcessInfo *pi = process_map->getptr(p_pid);
+	process_map_mutex.unlock();
+
+	if (pi && !pi->is_running) {
+		return false;
+	}
+
 	int status = 0;
 	if (waitpid(p_pid, &status, WNOHANG) != 0) {
+		if (pi) {
+			pi->is_running = false;
+			pi->exit_code = status;
+		}
 		return false;
 	}
 
 	return true;
 }
 
+int OS_Unix::get_process_exit_code(const ProcessID &p_pid) const {
+	process_map_mutex.lock();
+	const ProcessInfo *pi = process_map->getptr(p_pid);
+	process_map_mutex.unlock();
```

Comment:
Not sure if pointer returned by map can't be reallocated, so I would use (for both this and `is_process_running`):

```suggestion
	MutexLock(process_map_mutex);
	const ProcessInfo *pi = process_map->getptr(p_pid);
```